### ISSUE TITLE
`sort-keys` should detect keys used by other keys

### ISSUE THREAD

**Issue by cdoublev** [eslint/eslint#12043]
**Tell us about your environment**

* **ESLint Version:** ^6.1.0
* **Node Version:** v12.7.0
* **npm Version:** 6.10.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    env: { es6: true },
    extends: ['eslint:recommended'],
    parserOptions: {
        ecmaFeatures: { impliedStrict: true },
        ecmaVersion: 2019,
        sourceType: 'module',
    },
    plugins: ['compat'],
    rules: {
        'sort-keys': 'warn',
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const greeting = ({ who = 'world', message = `Hello ${who}` } = {}) => message
```

```bash
eslint src
```

**What did you expect to happen?**

`sort-keys` should not warn about `who` being defined before `message`.

**What actually happened? Please include the actual, raw output from ESLint.**

`warning  Expected object keys to be in ascending order. 'who' should be before 'message'`

**Are you willing to submit a pull request to fix this bug?**

I'm not familiar with ESLint rule development, but I'm open to take care of this.

**Comment by mdjermanovic** [comment]
Are you sure that the rule reports this particular line?

`sort-keys` is not supposed to check object patterns, this would be a bug just because of that. I can't reproduce it though, and the code that avoids object patterns looks ok.

Off-topic, there are other rules that don't work well with object patterns and those rules are probably not even intended to work with object pattern properties, I'll submit an issue.

**Comment by platinumazure** [comment]
Hi @cdoublev, thanks for the issue.

I can't reproduce this. My attempt to reproduce can be found in [this demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50LWVudiBlczYgKi9cbi8qIGVzbGludCBzb3J0LWtleXM6IHdhcm4gKi9cblxuY29uc3QgZ3JlZXRpbmcgPSAoeyB3aG8gPSAnd29ybGQnLCBtZXNzYWdlID0gYEhlbGxvICR7d2hvfWAgfSA9IHt9KSA9PiBtZXNzYWdlOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTAsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsiaW1wbGllZFN0cmljdCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6e319fQ==).

If there is something wrong with my reproduction attempt, please let me know where I made a mistake. Thanks!

**Comment by cdoublev** [comment]
Damn, the original function has changed since I submitted this issue, and I can't reproduce it anymore. I tried for the last 30 minutes. I'm sure that `sort-keys` were reporting this scenario, as I tested it a few times back and forth. It was a React component returning an `<a>`nchor element and receiving a `target` prop used by an "auto created" href prop, ie. ``href = `#${target}` ``. I removed it because I had the ESLint warning, and because I feel like it's a bad practice.

Anyway, I'm really sorry for waisting your time. Thank you both for your answers.
